upgrade/util/util.go

```diff
@@ -149,6 +159,97 @@ func CreateOrUpdateLonghornVersionSetting(namespace string, lhClient *lhclientse
 	return nil
 }
 
+// IsUpgradePathSupported returns if the upgrade path from lhVersionBeforeUpgrade to meta.Version is supported.
+func IsUpgradePathSupported(namespace string, lhClient *lhclientset.Clientset) bool {
+	lhVersionBeforeUpgrade, err := GetCurrentLonghornVersion(namespace, lhClient)
+	if err != nil {
+		logrus.WithError(err).Warn("failed to get current longhorn version")
+		return false
+	}
+
+	if lhVersionBeforeUpgrade == "" {
+		return true
+	}
+
+	logrus.Infof("Checking if the upgrade path (%v to %v) is supported", lhVersionBeforeUpgrade, meta.Version)
+
+	if !semver.IsValid(meta.Version) {
+		logrus.Warnf("Upgrading version is not valid: %v", meta.Version)
+		return false
+	}
+
+	lhNewMajorVersion := semver.Major(meta.Version)
+	lhMajorVersionBeforeUpgrade := semver.Major(lhVersionBeforeUpgrade)
+
+	lhNewMajorVersionNum, lhNewMinorVersionNum, err := getMajorMinorInt(meta.Version)
+	if err != nil {
+		logrus.WithError(err).Warnf("failed to parse %v major/minor version", meta.Version)
+		return false
+	}
+
+	lhMajorVersionBeforeUpgradeNum, lhMinorVersionBeforeUpgradeNum, err := getMajorMinorInt(lhVersionBeforeUpgrade)
+	if err != nil {
+		logrus.WithError(err).Warnf("failed to parse %v major/minor version", lhVersionBeforeUpgrade)
+		return false
+	}
+
+	if semver.Compare(lhMajorVersionBeforeUpgrade, lhNewMajorVersion) > 0 {
+		logrus.Warnf("Downgrading is not supported, %v to %v", lhVersionBeforeUpgrade, meta.Version)
+		return false
+	}
+
+	if semver.Compare(lhMajorVersionBeforeUpgrade, lhNewMajorVersion) < 0 {
+		if (lhNewMajorVersionNum - lhMajorVersionBeforeUpgradeNum) > 1 {
+			return false
+		}
+		return lhMinorVersionBeforeUpgradeNum >= LonghornV1ToV2MinorVersionNum
+	}
+
+	if (lhNewMinorVersionNum-lhMinorVersionBeforeUpgradeNum) > 1 || (lhNewMinorVersionNum-lhMinorVersionBeforeUpgradeNum) < 0 {
```

Comment:
`(lhNewMinorVersionNum-lhMinorVersionBeforeUpgradeNum) < 0 ` this is a downgrade case for minor version downgrade but the major is the same.

Reply:
Removed this duplicated check statement. Thanks.